hail/src/main/scala/is/hail/annotations/CodeOrdering.scala

```diff
@@ -376,4 +397,14 @@ abstract class CodeOrdering {
 
     (xm || ym).mux(xm && ym, equivNonnull(xv, yv))
   }
+
+  // reverses the sense of the non-null comparison only
+  def reverse: CodeOrdering = new CodeOrdering () {
+    override def reverse: CodeOrdering = CodeOrdering.this
+    override type T = CodeOrdering.this.T
+    override type P = CodeOrdering.this.P
+
+    def compareNonnull(x: Code[T], y: Code[T]): Code[Int] =
```

Comment:
This needs to override all the `xNonnull` methods, not just `compare`.

Reply:
It looks like the other ones are all implemented in terms of `compareNonnull`, if I reverse the notion of compare, the others should also be reversed, no?

e.g.:
```
co.reverse.ltNonnull(x, y)
 = co.reverse.compareNonnull(x, y) < 0
 = co.compareNonNull(y, x) < 0
```
if `x < y`, we'll get -1, which is indeed less than zero. If `x == y`, we'll get 0 which is not less than zero. If `x > y`, we'll get 1 which is not less than zero.

Reply:
This isn't true for all of the comparisons, which can override the base definitions. (e.g. `PFloat32`) The base definition was put in as a convenience, since it's true for most types (but not all).

Reply:
This `CodeOrdering` is an anonymous class that subclasses the base `CodeOrdering`, so it won't inherit any methods from any specific `CodeOrdering` subclasses like those for 32-bit floats.

This will require a few more JIT iterations to optimize since it has to eliminate the call to `java.lang.Float.compare`. I suspect it's not a significant performance issue, but I'll add the flipped implementations of the other methods anyway.

Reply:
ah, I see. Can you explain what happens if I call PFloat32().codeOrdering(mb, PFloat32()).reverse? I'm not quite sure I follow.

Reply:
I guess, more specifically, what happens if I look at `CodeOrdering.this.ltNonnull` inside the scope of the `CodeOrdering` class that `reverse` creates?

Reply:
[deleted a couple of comments because I think I was confusing myself as to what you meant by that]

That's why we need to explicitly define `ltNonnull` and friends, though. I want this to happen:
```
val ord = PFloat32().codeOrdering(mb, PFlost32()).reverse
assert(ord.equivNonnull(-0.0, 0.0))
assert(!ord.compareNonnull(-0.0, 0.0) > 0)
```

Reply:
Ah, `java.lang.Float.compare` considers 0 and -0 distinct but `FCMPL` does not. It seems dangerous to have these default implementations. I've split them into two semantically distinct classes, one permitting the default overrides and the other requiring the subclass to implement them.